lib/Optimizer/Transforms/PhaseFolding.cpp

```diff
@@ -0,0 +1,656 @@
+      if (new_phase.vars.contains(var))
+        new_phase.vars.remove(var);
+      else
+        new_phase.vars.insert(var);
+    new_phase.isInverted = (p1.isInverted != p2.isInverted);
+    return new_phase;
+  }
+
+  /// @brief Returns a new phase equal to `p1` with the opposite inversion flag
+  static Phase invert(Phase &p1) {
+    auto new_phase = Phase();
+    for (auto var : p1.vars)
+      new_phase.vars.insert(var);
+    new_phase.isInverted = !p1.isInverted;
+    return new_phase;
+  }
+
+  void dump() {
+    llvm::outs() << "Phase: ";
+    if (isInverted)
+      llvm::outs() << "!";
+    llvm::outs() << "{";
+    auto first = true;
+    for (auto var : vars) {
+      if (!first)
+        llvm::outs() << " ^ ";
+      llvm::outs() << var->idx;
+      first = false;
+    }
+    llvm::outs() << "}\n";
+  }
+};
+
+class PhaseStorage {
+  SmallVector<Phase> phases;
+  SmallVector<quake::RzOp> rotations;
+  size_t numCombined = 0;
+
+  /// @brief Merges the rotation at prev_idx with rzop by adding their
+  /// rotation angles and overwriting rzop's angle with the new
+  /// angle. The old rotation is erased. We keep the latter rotation
+  /// to ensure that dynamic rotation angles (e.g., dependent on
+  /// measurement results) are indeed available, as earlier angles
+  /// will always be available later, but not vice-versa.
+  void combineRotations(size_t prev_idx, quake::RzOp rzop) {
+    auto old_rzop = rotations[prev_idx];
+    auto rot_arg1 = old_rzop.getOperand(0);
+    auto rot_arg2 = rzop.getOperand(0);
+    auto builder = OpBuilder(rzop);
+    auto new_rot_arg =
+        builder.create<arith::AddFOp>(rzop.getLoc(), rot_arg1, rot_arg2);
+    rzop->setOperand(0, new_rot_arg.getResult());
+    old_rzop.erase();
+    rotations[prev_idx] = rzop;
+    numCombined++;
+  }
+
+public:
+  /// @brief registers a new rotation op for the given phase
+  /// @returns true if the rotation was combined, false otherwise
+  bool addOrCombineRotationForPhase(quake::RzOp op, Phase phase) {
+    for (size_t i = 0; i < phases.size(); i++)
+      if (phases[i] == phase) {
+        combineRotations(i, op);
+        return true;
+      }
+
+    phases.push_back(phase);
+    rotations.push_back(op);
+    return false;
+  }
+
+  size_t getNumCombined() { return numCombined; }
+};
+
+class PhaseFoldingPass
+    : public cudaq::opt::impl::PhaseFoldingBase<PhaseFoldingPass> {
+  using PhaseFoldingBase::PhaseFoldingBase;
+
+  // Perform the phase folding optimization over a subcircuit
+  void doPhaseFolding(Subcircuit *subcircuit) {
+    SmallVector<PhaseVariable *> phase_vars;
+    SmallVector<Phase> current_phases;
+    PhaseStorage store;
+    size_t i = 0;
+    // Initialize the phases and phase variables for each qubit in the circuit,
+    // the initial phase contains only the phase variable for that qubit
+    for (auto ref : subcircuit->getRefs()) {
+      auto phase_idx = i++;
+      auto new_phase_var = new PhaseVariable(phase_idx, ref);
```

Comment:
Q: Where do we clean up this `PhaseVariable`?

I saw that we pushed it to `phase_vars`, perhaps for tracking and cleaning purposes?